### ISSUE TITLE
Generate Receiving Address without Connection to `neptune-core`

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/neptune-core#299]
Previously, we had a workflow that enabled users to generate a public receiving address purely by interacting with `neptune-cli` and without ever running `neptune-core`. This functionality seems to have been lost over the course of several refactors or PRs, probably related to indexed derivation.

If I recall correctly, this was the workflow:
 - `neptune-cli generate-wallet` reads the `wallet.dat` file, or generates a new one if non exists, *and outputs the zeroth generation receiving address*.
 - `neptune-cli export-seed-phrase` prints the 18 seed phrase words.
 - `neptune-cli import-seed-phrase` prompts the user to input the 18 seed phrase words and creates a new `wallet.dat` file, but only if this file does not already exists.

The italic part of the first bullet point seems to have disappeared. Which is okay if there is another command that does the same. Unfortunately, `neptune-cli own-receiving-address` will *only* output a public receiving address if it can connect to a running instance of `neptune-core`.

In my opinion, this requirement is an unnecessary complication. It is particularly frustrating because we want premine beneficiaries to generate key pairs locally and send the public receiving address to us, and we want to lower the barrier for that task.

That said, there is a good reason why a CLI interface for displaying receiving addresses would want to interact with the node: because addresses should be unique and the node uses and modifies a counter involved in its derivation.

To address these limitations I would suggest moving forward thusly:
 - Rename command `own-receiving-address` to `next-receiving-address` to signal that it generates a new one every time (which explains why a connection to `neptune-core` is needed).
 - Add a command `premine-receiving-address` that generates the receiving address associated with index 0, without establishing a connection to `neptune-core`. After launch we could consider removing this command or renaming it to `static-receiving-address`.

**Comment by dan-da** [comment]
> Previously, we had a workflow that enabled users to generate a public receiving address purely by interacting with `neptune-cli` and without ever running `neptune-core`. This functionality seems to have been lost over the course of several refactors or PRs, probably related to indexed derivation.

displaying the 0th key was removed in commit 4182f7d39e947c86d401e60c6b43d570a25c24b6 for symmetric key notifications.  #171 has this note:

> neptune-cli generate-wallet: remove display of first wallet addr as it should not be considered special/only.

I still agree with this rationale.  In general the code should not treat any address as more special than another, and should not train users to think that way.    Unfortunately, the code does in fact presently treat the 0th key as special/reserved in that  coinbase rewards are always paid to it.  This is done for perf reasons and to avoid large derivation gaps.  Also it is expected (though not required) that premine rewards are paid to it.

> The italic part of the first bullet point seems to have disappeared. Which is okay if there is another command that does the same. Unfortunately, `neptune-cli own-receiving-address` will _only_ output a public receiving address if it can connect to a running instance of `neptune-core`.

> In my opinion, this requirement is an unnecessary complication.

It is only "unnecessary" for the 0th key.  Any subsequent key requires a connection to neptune core in order to derive the next key.  Note that once #304 is merged, it would be *possible* for neptune-cli to derive the nth key without a connection, however for `next-receiving-address` command it must know the last-known index, which resides in the wallet-database.    If neptune-cli could ever read/write that database directly, then it wouldn't need to ask neptune-core.  I don't think we want to actually do that, but it could be a future possibility.

> That said, there is a good reason why a CLI interface for displaying receiving addresses would want to interact with the node: because addresses should be unique and the node uses and modifies a counter involved in its derivation.

yep, yep.

> To address these limitations I would suggest moving forward thusly:
> 
> * Rename command `own-receiving-address` to `next-receiving-address` to signal that it generates a new one every time (which explains why a connection to `neptune-core` is needed).

agreed.  in fact, this was already done in #185, but didn't get merged somehow.    sigh...

> * Add a command `premine-receiving-address` that generates the receiving address associated with index 0, without establishing a connection to `neptune-core`. After launch we could consider removing this command or renaming it to `static-receiving-address`.

I propose that we make a pair of more general commands for offline usage: 
  `nth-spending-key <key-type> <n>`
  `nth-receiving-address <key-type> <n>`

notes:
 * allows any key to be derived, from index 0 to DerivationIndex::MAX.
 * supports both spending keys and receiving addresses.
 * supports all key-types, present and future.
 * key-type could default to 'generation'.
 * requires self-deriving keys, ie #304.

The proposed `premine-receiving-address` then maps to command `nth-receiving-address 0`.  I am ambivalent as to whether we create a separate/temporary command/alias for this purpose, or just provide premine recipients with instructions to use `nth-receiving-address 0`.

**Comment by dan-da** [comment]
One additional thought:

There has been up until now an expectation that premine rewards would be paid to the 0th key.  I do not see why that should necessarily be the case, or why a premine reward should be treated any differently from *any other tx* with regards to key generation.

In other words, what is wrong with instructing premine recipients something like:

1. start neptune-core if not already running by executing `neptune-core`.  
2. run command `neptune-cli next-receiving-address`.
3. send this address to neptune developers.

Personally, I think that is a perfectly reasonable instruction/requirement, and it teaches participants how to receive *any tx* rather than teaching a one-time special case -- basically an anti-pattern for providing address to sender.

I perceive there is a desire to not require a running neptune-core instance, so things are *as easy as possible* and perhaps to lessen support requests.   But I think that concern may be overstated/unjustified.  If our users can deal with a CLI, they can start neptune-core.  If they can't.... this is all too much for them anyway.

all that said, I am happy to add the nth_* commands once self-deriving keys are merged, as I think they are generally useful anyway.

**Comment by aszepieniec** [comment]
> The proposed `premine-receiving-address` then maps to command `nth-receiving-address 0`. I am ambivalent as to whether we create a separate/temporary command/alias for this purpose, or just provide premine recipients with instructions to use `nth-receiving-address 0`.

I am in favor of having the command `nth-receiving-address n` and of mapping `premine-receiving-address` to `nth-receiving-address 0`.

Premine recipients might want to verify before launch that their receiving address made it into the genesis block. They should be able to Ctrl-F in the code for their address, and reproduce the same address for this search. It cannot hurt to avoid confusion about which number to start counting from.

The point about supporting the anti-pattern is well received and based on this argument I *would* like to drop `premine-receiving-address` after launch.

**Comment by dan-da** [comment]
adding mainnet tag, as the premine-receiving-address seems necessary.  In fact, it should probably be in betanet-v2.
